python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -259,6 +267,42 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    def _detect_version_changes(self):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Saves this information in changed_versions in the Device instance.
+        """
+        old_device = self.get_existing_model()
+        self.changed_versions = {}
```

Comment:
PyLint would actually complain about this if it were still part of our CI pipeline: It's good practice to only create instance attributes in `__init__()`, rather than to create them dynamically at arbitrary times of the object's lifetime.

(It makes it a lot easier to discern what the valid attributes of an instance are if they are all defined for the first time in the same method)